src/integration-test/java/com/commercetools/sync/integration/commons/utils/CategoryITUtils.java

```diff
@@ -0,0 +1,345 @@
+     * "label": {
+     * "en": "backgroundColor"
+     * },
+     * "required": false,
+     * "type": {
+     * "name": "LocalizedString"
+     * },
+     * "inputHint": "SingleLine"
+     * },
+     * {
+     * "name": "invisibleInShop",
+     * "label": {
+     * "en": "invisibleInShop"
+     * },
+     * "required": false,
+     * "type": {
+     * "name": "Boolean"
+     * },
+     * "inputHint": "SingleLine"
+     * }
+     * ]
+     *
+     * @param locale defines the locale for which the field definition names are going to be bound to.
+     * @return the list of field definitions.
+     */
+    private static List<FieldDefinition> buildCategoryCustomTypeFieldDefinitions(@Nonnull final Locale locale) {
+        return asList(
+            FieldDefinition
+                .of(LocalizedStringFieldType.of(), LOCALISED_STRING_CUSTOM_FIELD_NAME,
+                    LocalizedString.of(locale, LOCALISED_STRING_CUSTOM_FIELD_NAME), false),
+            FieldDefinition
+                .of(BooleanFieldType.of(), BOOLEAN_CUSTOM_FIELD_NAME,
+                    LocalizedString.of(locale, BOOLEAN_CUSTOM_FIELD_NAME), false));
+
+    }
+
+
+    /**
+     * Deletes up to {@link SphereClientUtils#QUERY_MAX_LIMIT} categories from CTP
+     * projects defined by the {@code ctpClient}. Only issues delete request action to a category in the case that none
+     * of its ancestors was already deleted or not to avoid trying to delete a category which would have been already
+     * deleted, due to deletion of an ancestor of it. As a performance improvement, this method sorts categories by
+     * least ancestors for faster deletion (due to deletion ancestors always first, which in turn deletes all the
+     * children and grand children.
+     *
+     *
+     * @param ctpClient defines the CTP project to delete the categories from.
+     */
+    public static void deleteAllCategories(@Nonnull final SphereClient ctpClient) {
+        final Set<String> keys = new HashSet<>();
+        final List<Category> categories = QueryExecutionUtils.queryAll(ctpClient,
+            CategoryQuery.of().withExpansionPaths(CategoryExpansionModel::ancestors))
+                                                             .thenApply(CategoryITUtils::sortCategoriesByLeastAncestors)
+                                                             .toCompletableFuture().join();
+        categories.forEach(category -> {
+            final String categoryKey = category.getKey();
+            if (!hasADeletedAncestor(category, keys)) {
+                ctpClient.execute(CategoryDeleteCommand.of(category))
+                         .thenAccept(deletedCategory -> keys.add(categoryKey))
+                         .toCompletableFuture().join();
+            }
+        });
+    }
+
+    private static List<Category> sortCategoriesByLeastAncestors(@Nonnull final List<Category> categories) {
+        categories.sort(Comparator.comparingInt(category -> category.getAncestors().size()));
```

Comment:
Maybe it's a bit SDK related question but does a root category have an ancestor too? I see that it is not nullable. Or is it an empty list of references when it is root?

Reply:
exactly you should get an empty list if it doesn't have ancestors.